src/katgpucbf/fgpu/kernels/ddc.mako

```diff
@@ -14,307 +14,234 @@
- * - SUBSAMPLING must be a multiple of SG_SIZE
- * - COARSEN should be odd for best performance
- * - SG_SIZE should be a power of two for best performance
+ * - C * SUBSAMPLING * INPUT_SAMPLE_BITS must be a multiple of SAMPLE_WORD_BITS
  */
 
+<%include file="/port.mako"/>
+
 #define WGS ${wgs}
 #define TAPS ${taps}
 #define SUBSAMPLING ${subsampling}
-#define COARSEN ${coarsen}
-#define SG_SIZE ${sg_size}
-#define SAMPLE_BITS ${sample_bits}
-#define SEGMENT_SAMPLES ${segment_samples}
+#define INPUT_SAMPLE_BITS ${input_sample_bits}
+#define C ${unroll}
+#define W ${(taps + subsampling - 1) // subsampling}
 
-// Number of contiguous 32-bit words to store in each segment
-#define SEGMENT_WORDS (SEGMENT_SAMPLES * SAMPLE_BITS / 32)
-// Number of output samples
-#define GROUP_OUT_SIZE (COARSEN * (WGS / SG_SIZE))
-// Stride of input samples between workgroups
-#define GROUP_IN_SIZE (GROUP_OUT_SIZE * SUBSAMPLING)
-// Number of input samples to load in this workgroup
-#define LOAD_SIZE (GROUP_IN_SIZE + TAPS - SUBSAMPLING)
-// Load-size expressed in 32-bit words (rounding up)
-#define LOAD_WORDS ((LOAD_SIZE - 1) * SAMPLE_BITS / 32 + 1)
-// Number of segments to load per work-item
-#define SEGMENTS ((LOAD_SIZE - 1) / (SEGMENT_SAMPLES * WGS) + 1)
-/* Number of contiguous samples that take turns occupying a tile
- * (must divide both SEGMENT_SAMPLES and SUBSAMPLING, and be a multiple
- * of SG_SIZE). This implementation is pessimistic when SUBSAMPLING is
- * neither a factor nor multiple of SEGMENT_SAMPLES, but that's not expected to
- * be a common case.
- */
-#if SUBSAMPLING % SEGMENT_SAMPLES == 0
-# define TILE_SAMPLES SEGMENT_SAMPLES
-#elif SEGMENT_SAMPLES % SUBSAMPLING == 0
-# define TILE_SAMPLES SUBSAMPLING
-#else
-# define TILE_SAMPLES SG_SIZE
-#endif
-// Number of tiles to store in local memory
-#define TILES (LOAD_SIZE / TILE_SAMPLES)
-#define TILES_PER_SEGMENT (SEGMENT_SAMPLES / TILE_SAMPLES)
-#define TILES_PER_SUBSAMPLING (SUBSAMPLING / TILE_SAMPLES)
+/// Raw storage type for sample data
+typedef unsigned int sample_word;
+/// Signed version of sample_word
+typedef int ssample_word;
+#define SAMPLE_WORD_BITS 32
 
-${wg_reduce.define_scratch('float', sg_size, 'scratch_t', allow_shuffle=True)}
-${wg_reduce.define_function('float', sg_size, 'reduce', 'scratch_t', wg_reduce.op_plus, allow_shuffle=True, broadcast=False)}
+DEVICE_FN static unsigned int reverse_endian(unsigned int v)
+{
+    return __byte_perm(v, v, 0x0123);
+}
 
-DEVICE_FN void sync()
+/**
+ * Load the next sample value.
+ *
+ * This is intended to be called with a sequence of contiguous sample indices,
+ * passing the same @a buffer each time. On the first call with a particular
+ * buffer (or after a discontiguous change in @a idx), set @a start to false.
+ *
+ * @param in      Array with all the raw (but native-endian) sample words for the workgroup
+ * @param buffer  Sample word that help the LSBs of the previous sample if any (updated on return)
```

Comment:
```suggestion
 * @param buffer  Sample word that helps the LSBs of the previous sample if any (updated on return)
```
?

Though the entire sentence doesn't quite make sense to me, maybe I've just missed some context.